internal/service/managedobjectstorage/managed_object_storage.go

```diff
@@ -0,0 +1,360 @@
+package managedobjectstorage
+
+import (
+	"context"
+	"fmt"
+	"regexp"
+	"time"
+
+	"github.com/UpCloudLtd/terraform-provider-upcloud/internal/utils"
+
+	"github.com/UpCloudLtd/upcloud-go-api/v6/upcloud"
+	"github.com/UpCloudLtd/upcloud-go-api/v6/upcloud/request"
+	"github.com/UpCloudLtd/upcloud-go-api/v6/upcloud/service"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
+)
+
+func ResourceManagedObjectStorage() *schema.Resource {
+	return &schema.Resource{
+		Description:   "This resource represents an UpCloud Managed Object Storage instance, which provides S3 compatible storage.",
+		CreateContext: resourceManagedObjectStorageCreate,
+		ReadContext:   resourceManagedObjectStorageRead,
+		UpdateContext: resourceManagedObjectStorageUpdate,
+		DeleteContext: resourceManagedObjectStorageDelete,
+		Importer: &schema.ResourceImporter{
+			StateContext: schema.ImportStatePassthroughContext,
+		},
+		Schema: map[string]*schema.Schema{
+			"configured_status": {
+				Description: "Service status managed by the end user.",
+				Required:    true,
+				Type:        schema.TypeString,
+				ValidateFunc: validation.StringInSlice(
+					[]string{
+						string(upcloud.ManagedObjectStorageConfiguredStatusStarted),
+						string(upcloud.ManagedObjectStorageConfiguredStatusStopped),
+					},
+					false,
+				),
+			},
+			"created_at": {
+				Description: "Creation time.",
+				Computed:    true,
+				Type:        schema.TypeString,
+			},
+			"endpoint": {
+				Description: "Endpoints for accessing the Managed Object Storage service.",
+				Computed:    true,
+				Type:        schema.TypeSet,
+				Elem:        schemaEndpoint(),
+			},
+			"labels": utils.LabelsSchema("network gateway"),
+			"network": {
+				Description: "Attached networks from where object storage can be used. Private networks must reside in object storage region.",
+				Optional:    true,
+				Type:        schema.TypeSet,
+				Elem:        schemaNetwork(),
+			},
+			"operational_state": {
+				Description: "Operational state of the Managed Object Storage service.",
+				Computed:    true,
+				Type:        schema.TypeString,
+			},
+			"region": {
+				Description: "Region in which the service will be hosted, see `upcloud_object_storage2_regions` data source.",
+				Required:    true,
+				Type:        schema.TypeString,
+			},
+			"updated_at": {
+				Description: "Creation time.",
+				Computed:    true,
+				Type:        schema.TypeString,
+			},
+			"users": {
+				Description: "List of cloud Users allowed to use object storage. See `upcloud_managed_object_storage_user_access_key` for managing access keys.",
```

Comment:
Does _cloud user_ mean main-account and/or sub-account here or is this related to a DB resource?

Reply:
Current main account and it's sub-accounts